src/app/games/hive/HiveState.ts

```diff
@@ -0,0 +1,211 @@
+        });
+        return remaining;
+    }
+
+    public getAllRemaining(player: Player): HivePiece[] {
+        const remaining: HivePiece[] = [];
+        this.pieces.forEach((item: {key: HivePiece, value: number}) => {
+            if (item.key.owner === player && item.value > 0) remaining.push(item.key);
+        });
+        return remaining;
+    }
+}
+
+export class HiveState extends FreeHexagonalGameState<HivePieceStack> implements ComparableObject {
+
+    public readonly width: number;
+
+    public readonly height: number;
+
+    public static getInitialState(): HiveState {
+        const board: Table<HivePiece[]> = [];
+        return HiveState.fromRepresentation(board, 0);
+    }
+
+    public static fromRepresentation(board: Table<HivePiece[]>, turn: number): HiveState {
+        const pieces: ReversibleMap<Coord, HivePieceStack> = new ReversibleMap<Coord, HivePieceStack>();
+        const remainingPieces: HiveRemainingPieces = HiveRemainingPieces.getInitial();
+        const queenBees: MGPMap<Player, Coord> = new MGPMap();
+        for (let y: number = 0; y < board.length; y++) {
+            for (let x: number = 0; x < board[0].length; x++) {
+                if (board[y][x].length > 0) {
+                    pieces.set(new Coord(x, y), new HivePieceStack(board[y][x]));
+                    const queenBee: MGPOptional<HivePiece> =
+                        MGPOptional.ofNullable(board[y][x].find((piece: HivePiece) => piece.kind === 'QueenBee'));
+                    if (queenBee.isPresent()) {
+                        queenBees.set(queenBee.get().owner, new Coord(x, y));
+                    }
+                    for (const piece of board[y][x]) {
+                        remainingPieces.remove(piece);
+                    }
+                }
+            }
+        }
+        return new HiveState(pieces, remainingPieces, queenBees, turn);
+    }
+
+    public constructor(pieces: ReversibleMap<Coord, HivePieceStack>,
+                       public readonly remainingPieces: HiveRemainingPieces,
+                       private readonly queenBees: MGPMap<Player, Coord>,
+                       turn: number)
+    {
+        super(pieces, turn);
+        for (const player of queenBees.listKeys()) {
+            const oldCoord: Coord = queenBees.get(player).get();
+            const newCoord: Coord = oldCoord.getNext(this.offset);
+            queenBees.replace(player, newCoord);
+        }
+    }
+
+    public equals(other: HiveState): boolean {
+        return this.pieces.equals(other.pieces) &&
+            this.remainingPieces.equals(other.remainingPieces) &&
+            this.queenBees.equals(other.queenBees) &&
+            this.turn === other.turn;
+    }
+
+    public getCopy(): HiveState {
+        return new HiveState(this.pieces.getCopy(),
+                             this.remainingPieces.getCopy(),
+                             this.queenBees.getCopy(),
+                             this.turn);
+    }
```

Comment:
kill the baloon, kill the baloon !
(en vrai je propose de butter cette fonction car elle déplaît a notre seigneurerie le pattern d'immutabilité)

Reply:
elle est morte ! Après la v1 on réimplémente MGP en haskell !